### ISSUE TITLE
support for multi-label classification

### ISSUE THREAD

**Issue by xdtl** [fizyr/keras-retinanet#1095]
As I understand, the current focal loss for classification is sigmoid+binary_crossentropy, which is able to be extended to support multi-label classification. I tried to modify function 'anchor_targets_bbox' to convert those one-hot vectors to 'multi-hot vector'. I wonder if that is the only change I need to make. Any suggestions on how to extend it to support multi-label classification are greatly appreciated!

In addition, the initial classification loss is supposed to be the negative log probability of the correct class, so in my case of 3 classes+bg: -ln(1/4) = 1.4. But my initial loss is 3.x, much larger than 1.4. Does that mean something is wrong in classification? Thanks!

**Comment by hgaiser** [comment]
I think indeed if you modify the generator you should get multi-label classification to work.

> In addition, the initial classification loss is supposed to be the negative log probability of the correct class, so in my case of 3 classes+bg: -ln(1/4) = 1.4. But my initial loss is 3.x, much larger than 1.4. Does that mean something is wrong in classification? Thanks!

Is that calculation correct? You're not taking focal loss into account I believe.

**Comment by xdtl** [comment]
I see... Thanks for your reply!!

**Comment by 0chandansharma** [comment]
@xdtl, were you able to convert code for multi-label detection?

**Comment by xdtl** [comment]
I tried that, but it didn't generate satisfactory results. Actually when creating training labels, I think it is okay to have something like this:
/data/imgs/img_001.jpg,837,346,981,456,label1
/data/imgs/img_001.jpg,837,346,981,456,label2
Then during inference, if I find multiple detections with significant overlap, I can output them as one object with multi-label. Hope it helps.

**Comment by hgaiser** [comment]
> I tried that, but it didn't generate satisfactory results. Actually when creating training labels, I think it is okay to have something like this:
> /data/imgs/img_001.jpg,837,346,981,456,label1
> /data/imgs/img_001.jpg,837,346,981,456,label2
> Then during inference, if I find multiple detections with significant overlap, I can output them as one object with multi-label. Hope it helps.

This will create two separate objects, which is not what you want. You want one object with multiple classes.

**Comment by xdtl** [comment]
Yes, they are separate objects from the output of retinanet. Then in a post-processing step, if I find two objects with the same location, I take them as one object with multiple classes.

**Comment by hgaiser** [comment]
That sounds like a workaround for a problem that you introduced to be honest.

**Comment by xdtl** [comment]
Sorry for the confusion. My point is Retinanet can be used as it is to deal with the multi-label classification problem. No change is needed to the network itself. Since ground truth bounding box can overlap partially, they can also overlap entirely.

**Comment by 0chandansharma** [comment]
So, @hgaiser  can you help me to conclude about the multi-label classification using Keras-Retinanet.
Hope us understand the problem.
I mean it is possible or not with the existing architecture? , or only the solution, which is given by @xdtl.
